packages/smithy-aws-core/src/smithy_aws_core/user_agent.py

```diff
@@ -0,0 +1,269 @@
+            return [UserAgentComponent("md", "arch", self._platform_machine.lower())]
+        return []
+
+    def _build_language_metadata(self) -> List[UserAgentComponent]:
+        """Build the language components of the User-Agent header string.
+
+        Returns the Python version in a component with prefix "lang" and name
+        "python". The Python implementation (e.g. CPython, PyPy) is returned as
+        separate metadata component with prefix "md" and name "pyimpl".
+
+        String representation of an example return value:
+        ``lang/python#3.10.4 md/pyimpl#CPython``
+        """
+        lang_md = [
+            UserAgentComponent("lang", "python", self._python_version),
+        ]
+        if self._python_implementation:
+            lang_md.append(
+                UserAgentComponent("md", "pyimpl", self._python_implementation)
+            )
+        return lang_md
+
+    def _build_execution_env_metadata(self) -> List[UserAgentComponent]:
+        """Build the execution environment component of the User-Agent header.
+
+        Returns a single component prefixed with "exec-env", usually sourced from the
+        environment variable AWS_EXECUTION_ENV.
+        """
+        if self._execution_env:
+            return [UserAgentComponent("exec-env", self._execution_env)]
+        else:
+            return []
+
+    def _build_feature_metadata(self) -> List[UserAgentComponent]:
+        """Build the features components of the User-Agent header string.
+
+        TODO: These should be sourced from property bag set on context.
+        """
+        return []
+
+    def _build_app_id(self) -> List[UserAgentComponent]:
+        """Build app component of the User-Agent header string."""
+        if self._user_agent_app_id:
+            return [UserAgentComponent("app", self._user_agent_app_id)]
+        else:
+            return []
+
+    def _build_suffix(self) -> List[_UAComponent]:
+        if self._user_agent_suffix:
+            return [RawStringUserAgentComponent(self._user_agent_suffix)]
+        else:
+            return []
+
+
+def sanitize_user_agent_string_component(raw_str: str, allow_hash: bool = False) -> str:
+    """Replaces all not allowed characters in the string with a dash ("-").
+
+    Allowed characters are ASCII alphanumerics and ``!$%&'*+-.^_`|~``. If
+    ``allow_hash`` is ``True``, "#"``" is also allowed.
+
+    :type raw_str: str
+    :param raw_str: The input string to be sanitized.
+
+    :type allow_hash: bool
+    :param allow_hash: Whether "#" is considered an allowed character.
+    """
+    return "".join(
+        c if c in _USERAGENT_ALLOWED_CHARACTERS or (allow_hash and c == "#") else "-"
+        for c in raw_str
+    )
+
+
+def _get_crt_version() -> str | None:
+    """This function is considered private and is subject to abrupt breaking changes."""
+    try:
+        import awscrt
+
+        return awscrt.__version__
+    except AttributeError:
+        return None
```

Comment:
I'm wondering if we fold the `HAS_CRT` check into this call so it does the None/"Unknown" work for us.